### PR TITLE
boot: Read files in small chunks on broken firmware

### PR DESCRIPTION
**Pull request by medhefgo**
Fixes: #25911

### CODE REVIEW

File: src/boot/efi/util.c [hunk 1]
```diff
@@ -293,6 +293,59 @@ void mangle_stub_cmdline(char16_t *cmdline) {
         }
 }
 
+EFI_STATUS chunked_read(EFI_FILE *file, size_t *size, void *buf) {
+        EFI_STATUS err;
+
+        assert(file);
+        assert(buf);
```
Comment (poettering) [inline src/boot/efi/util.c:290]: assert(size)

File: src/boot/efi/util.c [hunk 2]
```diff
@@ -293,6 +293,59 @@ void mangle_stub_cmdline(char16_t *cmdline) {
...
+
+        /* This is a drop-in replacement for EFI_FILE->Read() with the same API behavior.
+         * Some broken firmwares cannot handle large file reads and will instead return
+         * an error. As a workaround, read such files in small chunks. */
```
Comment (poettering) [inline src/boot/efi/util.c]: maybe reference the issue this was reported in here

**Review by poettering — COMMENTED** [review]
lgtm, if this works.

just some nitpicks

**Comment by poettering** [comment]
@kernle32dll please report back if this makes things work for you

**Comment by kernle32dll** [comment]
@poettering 

As mentioned in https://github.com/systemd/systemd/issues/25911, this is now confirmed to work for me.

**Comment by flokli** [comment]
Thanks! Can this also be backported to systemd-stable? https://github.com/NixOS/nixpkgs/issues/227431#issuecomment-1556474041 mentions this also systemd 253.

**Comment by medhefgo** [comment]
Yes, should be fine to backport.

**Comment by aderumier** [comment]
Hi, could be great to have a backport to systemd-stable 252, for debian

https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1039508

**Comment by antongisli** [comment]
Backporting this would be _amazing_ because on e.g. Ubuntu (24.04) there is no chance to get them updated to a later version. It would open up this technology for a much wider subsection of OS versions and help get adoption. We're using this with Kairos for Trusted Boot and with end users who need to use e.g. Ubuntu it is putting us in a tough spot.

**Comment by bluca** [comment]
This is already in Ubuntu 24.04

**Comment by antongisli** [comment]
thank you @bluca! I see that now, nice!

**Pull request merged**